core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -212,15 +212,26 @@ public final Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEnt
 
    private Object visitDataReadCommand(InvocationContext ctx, AbstractDataCommand command) {
       final Object key = command.getKey();
-      entryFactory.wrapEntryForReading(ctx, key, command.getSegment(), ignoreOwnership(command) || canRead(command));
-      return invokeNextThenApply(ctx, command, dataReadReturnHandler);
+      CompletionStage<Void> stage = entryFactory.wrapEntryForReading(ctx, key, command.getSegment(),
+            ignoreOwnership(command) || canRead(command));
+      if ((CompletionStages.isCompletedSuccessfully(stage))) {
+         return invokeNextThenApply(ctx, command, dataReadReturnHandler);
+      }
+      return asyncValue(stage).thenApply(ctx, command, ((rCtx, rCommand, rv) ->
+         invokeNextThenApply(rCtx, command, dataReadReturnHandler)));
    }
 
    @Override
    public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) throws Throwable {
       boolean ignoreOwnership = ignoreOwnership(command);
+      AggregateCompletionStage<Void> aggregateCompletionStage = null;
       for (Object key : command.getKeys()) {
-         entryFactory.wrapEntryForReading(ctx, key, keyPartitioner.getSegment(key), ignoreOwnership || canReadKey(key));
+         CompletionStage<Void> stage = entryFactory.wrapEntryForReading(ctx, key, keyPartitioner.getSegment(key),
+               ignoreOwnership || canReadKey(key));
+         aggregateCompletionStage = accumulateStage(stage, aggregateCompletionStage);
+      }
+      if (aggregateCompletionStage != null) {
+
```

Comment:
isn't something missing here? maybe wait for the stages to complete?

Reply:
ah yes, I guess I forgot to put a TODO when I left it and didn't get back to fixing it :)